### DISCUSSION TITLE
V3 How do we connect two nodes?

### DISCUSSION THREAD

**Discussion by davidroman0O** [ergo-services/ergo#177] [Q&A]
I've been looking at the v3 codebase and played with the ergo_examples repo on the v3 branch, there are lot of examples on how to setup one node with all the tooling `ergo` is providing (udp, tcp, http, web server, etc), so far so good!

I know you're going to release the documentation soon but in the same time I would like to continue playing with it!

I saw there is a "registrar" that is supposed to help to connect nodes, I'm a little bit confused, I tried but i end up with `err` being filled for various reasons, I'm pretty sure i'm missing one little thing!

Do you have a simple code snippet of a simple ping-pong between two different runtimes? 🙏

**Comment by davidroman0O** [comment]
Just posting what I had, to test that ping-pong, what `not supported` means in that case? Do I miss a hierarchy permission thing? 


ping 

```go
package main

import (
	"fmt"
	"log"
	"time"

	"ergo.services/ergo"
	"ergo.services/ergo/act"
	"ergo.services/ergo/gen"
)

type Ping struct {
	act.Actor
}

func (p *Ping) HandleCall(from gen.PID, ref gen.Ref, request any) (any, error) {
	switch request.(type) {
	case string:
		log.Println("Ping received pong, sending ping")
		return "ping", nil
	}
	return nil, fmt.Errorf("unexpected request: %#v", request)
}

func (p *Ping) HandleMessage(from gen.PID, message any) error {
	switch message.(type) {
	case string:
		log.Println("Ping received pong, sending ping")
		_, err := p.Call(from, "ping")
		return err
	}
	return fmt.Errorf("unexpected message: %#v", message)
}

func NewPing() gen.ProcessBehavior {
	return &Ping{}
}

func main() {
	node, err := ergo.StartNode("ping@localhost", gen.NodeOptions{
		Network: gen.NetworkOptions{
			Cookie:             "cookie",
			InsecureSkipVerify: true,
		},
	})
	if err != nil {
		log.Fatalf("Failed to start ping: %v", err)
	}
	defer node.Stop()

	node.Spawn(NewPing, gen.ProcessOptions{})

	go func() {
		time.Sleep(1 * time.Second)
		if err = node.Send("pong@localhost", "ping"); err != nil {
			fmt.Printf("Unable to send message to node '%s': %s\n", "pong@localhost", err)
			return
		}
	}()

	node.Wait()
}


```


pong 

```go
package main

import (
	"fmt"
	"log"

	"ergo.services/ergo"
	"ergo.services/ergo/act"
	"ergo.services/ergo/gen"
)

type Pong struct {
	act.Actor
}

func (p *Pong) HandleCall(from gen.PID, ref gen.Ref, request any) (any, error) {
	switch request.(type) {
	case string:
		log.Println("Pong received ping, sending pong")
		return "pong", nil
	}
	return nil, fmt.Errorf("unexpected request: %#v", request)
}

func (p *Pong) HandleMessage(from gen.PID, message any) error {
	switch message.(type) {
	case string:
		log.Println("Pong received ping, sending pong")
		_, err := p.Call(from, "pong")
		return err
	}
	return fmt.Errorf("unexpected message: %#v", message)
}

func NewPong() gen.ProcessBehavior {
	return &Pong{}
}

func main() {
	node, err := ergo.StartNode("pong@localhost", gen.NodeOptions{
		Network: gen.NetworkOptions{
			Cookie:             "cookie",
			InsecureSkipVerify: true,
		},
	})
	if err != nil {
		log.Fatalf("Failed to start pong: %v", err)
	}
	defer node.Stop()

	node.Spawn(NewPong, gen.ProcessOptions{})

	node.Wait()
}


```

**Comment by halturin** [comment]
Things that need to be fixed in your example:
1. Use `SpawnRegister` to be able to send messages by the process name. for example 
    `node.SpawnRegister("pong_process", NewPong, gen.ProcessOptions{})`
3. Incorrect process name "pong@localhost". Use `gen.ProcessID` for sending messages to the remote process by name
    `node.Send(gen.ProcessID{Name: "pong_process", Node: "pong@localhost"}, message)`
  that was the reason of your `"not supported"` error. Node couldn't recognize the destination in `node.Send("pong@localhost", "ping")`. Destination must be:
![image](https://github.com/ergo-services/ergo/assets/118860/81b08ed7-eb6b-483d-8446-1cd9299633b5)

    You can use `node.Send(gen.Atom("pong_process"), message)` only if that process was spawned on the same node.

Beside that...

  - You are sending messages using `Send(...)` method of `gen.Node` interface. It uses `CorePID` as a sender which is a virtual process identifier with no process under the hood (no mailbox for the messages, no handlers etc). 
  - You will get the `gen.ErrTimeout` in `p.Call(...)` since no real process with that pid.

Please, take a look at "call" example https://github.com/ergo-services/examples/tree/v300. It does exactly what you want:
![image](https://github.com/ergo-services/ergo/assets/118860/178947ca-79bd-492f-9a33-c9489a4a483f)

**Reply by davidroman0O** [reply]
Ooooh ok it works now, i see! 

Oh I completely missed that line `remote := gen.ProcessID{Name: "b", Node: "nodeB@localhost"}` on your example! 🤦‍♂️

Thank you for your answer, very instructive! I can't wait to read the doc!

Now I'm curious to explore what's that `alias` thing! 😄  
I will continue to read the code and explore until i have thing ive missed or don't understand

Thanks again!
